Add tags that are present next to instance name to ease sorting

stephengeller opened the pull request:
## What does this change?

It was highlighted by @rtyley that some AWS accounts have multiple teams using them, meaning multiple teams might have to sift through a potentially large list of instances to find out which ones apply to them.

This PR implements the suggested improvement to list the tags that _are_ present next to the instance name to ease sorting: 
![image](https://user-images.githubusercontent.com/25747336/99651453-84dfb080-2a4e-11eb-9d1c-dcd9a0a28b54.png)

This lambda also did not have tests, so this PR sets up a test suite and adds tests for this logic.

## How to test

Within the `lambda` directory, run `yarn test`.

## How can we measure success?

It is easier to see at a glance which instances belong to which app/stack/stage.

rtyley reviewed (commented):
This looks great to me! I've added your instructions about how to run the tests to the readme in additional commit 72eea4f0, I hope that's ok!

One other thing that struck me, after my email yesterday, was that the EC2 instance type (eg `t3.medium`) is also often helpful in identifying responsibility when no tags at all are present - different teams can tend to use different instance types (eg the Data Lake team often use memory-optimised r4.2xlarge instances, which we on the Ophan team don't!). I think that information is available in the code:

https://github.com/guardian/tag-janitor/blob/08fa58553e41c3eec2285b34e1ff029fa252825e/lambda/src/interfaces.ts#L30

...so it might be a fairly easy thing to add, though I guess whether it might be more noisy than useful is debatable.

akash1810 commented on README.adoc:
```diff
@@ -32,6 +32,17 @@ $ yarn start # runs the handler locally
+
+[source,bash]
+----
+$ yarn add --dev jest
```
I don't think we need to explicitly install Jest as it is listed as a `devDependency` in package.json. So this would become "run `yarn test`".

stephengeller replied:
I think you're right! I've updated this to instruct to just install all dependencies first.

rtyley replied:
@akash1810 @stephengeller  As it happened, when I ran `yarn test` on my machine, it did fail, and I _did_ have to execute `yarn add --dev jest` to get it to work.

...edit, oh, I see, executing just `yarn` by itself takes the place of doing `yarn add --dev jest` - and executing `yarn test` isn't enough to get the dependencies install.

akash1810 commented on lambda/test/handler.test.ts:
```diff
@@ -0,0 +1,88 @@
+          instanceName: "instance2",
+          meta: { href: "href2" },
+        },
+      ] as unknown) as Instance[]);
```
Ooh this is interesting! Could we add a comment against this casting?

stephengeller replied:
Sounds good! I'm struggling to come up with a good comment though, do you have an example? Is this even the best approach for this?

akash1810 replied:
The `Partial<Instance>` specification looks like a better solution than `as unknown as Instance[]`; it looks like there are a couple more instances of `as unknown` that we can replace with this too.

stephengeller replied:
Thanks! Didn't realise I missed the other ones, they've now been updated too.

akash1810 reviewed (commented):
This is great! Added a couple of minor comments.

In addition to reporting the instance type as suggested by @rtyley , it would be great to report which cloudformation stack the instance belonged to too. This is dependant on Prism crawling cloudformation which it does not do currently, cc @Nirvikalpa108 @sihil .

stephengeller commented:
> This looks great to me! I've added your instructions about how to run the tests to the readme in additional commit [72eea4f](https://github.com/guardian/tag-janitor/commit/72eea4f0193e725ee29914f00e75be5826e4c0ed), I hope that's ok!
> 
> One other thing that struck me, after my email yesterday, was that the EC2 instance type (eg `t3.medium`) is also often helpful in identifying responsibility when no tags at all are present - different teams can tend to use different instance types (eg the Data Lake team often use memory-optimised r4.2xlarge instances, which we on the Ophan team don't!). I think that information is available in the code:
> 
> https://github.com/guardian/tag-janitor/blob/08fa58553e41c3eec2285b34e1ff029fa252825e/lambda/src/interfaces.ts#L30
> 
> ...so it might be a fairly easy thing to add, though I guess whether it might be more noisy than useful is debatable.

Thanks for the README addition @rtyley ! Looks great - I've taken Akash's comment and updated it as I think he's right in us not needing to manually install Jest.

I like this idea! Would this be in place of the `(App: foo, Stack: bar)` information being added in this PR? I would think it wouldn't be hard to configure it to show the instance type if **all** tags are missing from the instance. If so, that works and shouldn't be hard to implement.

rtyley commented:
>> ...the EC2 instance type (eg t3.medium) is also often helpful in identifying responsibility when no tags at all are present...

> I like this idea! Would this be in place of the `(App: foo, Stack: bar)` information being added in this PR? I would think it wouldn't be hard to configure it to show the instance type if **all** tags are missing from the instance. If so, that works and shouldn't be hard to implement.

Personally, I'd be happy with the instance type always being shown, regardless of the tags! And I wouldn't want to lose the `(App: foo, Stack: bar)` information, as I think it's much much meaningful!

stephengeller commented:
Okay great, sounds good! I agree about not losing the present tags info, let's try adding the instance type as well as the tags that _do_ exist in that parenthesis after the instance ID and see how it goes.

stephengeller commented:
This has been updated to add the instance type as per @rtyley 's suggestion. I've also created a helper fn in the test to allow us to generate `Instance` objects using `Partial<Instance>` which should hopefully make it easier to read as per @akash1810 's comment on the casting to `unknown`.

stephengeller commented:
Testing a comment to see if the chat bot has picked up this repo

rtyley reviewed (approved):
Looks good to me, look forward to seeing in this in the next email sent to us!
